### PR TITLE
'Be this camera' incremental improvements

### LINKED ISSUE #13525: Bug Report: Camera editor selection and other cameras move to the wrong position
**Describe the bug**
switching to other cameras the wrong position

**Screenshots/Video**

https://user-images.githubusercontent.com/17591539/205438132-14bc2c61-f242-4302-8aa1-ef893794ccb3.mp4

https://user-images.githubusercontent.com/17591539/205438135-d2cb7ea7-a0d3-4b70-a5f7-53c806a10b01.mp4

**Found in Branch**
development

**Commit ID from [o3de/o3de](https://github.com/o3de/o3de) Repository**
4ac8adffbe54068a43b7e90d3248b7cf309ee843

**Desktop/Device (please complete the following information):**
 - Device: PC
 - OS: Fedora
 - Version 36

### LINKED ISSUE #13763: "Be This Camera" Mode Blocks Editing Transform
**Describe the bug**
While in "be this camera mode" I cannot make micro-adjustments to the camera position by changing the values by hand in the transform component.

**Steps to reproduce**
1. Open the default level
3. Select the camera entity
4. Select "Be this camera"
5. Try to edit the camera position by editing the transform component

**Expected behavior**
Editing the transform component should move the camera

**Actual behavior**
User must click "Return to default editor camera" in order to hand-edit the transform 

**Screenshots/Video**
image

Notice in the video I change the X rotation value to 0.0, 2.0, and 10.0 but it always snaps back to the original value.
https://user-images.githubusercontent.com/32776221/208368292-00d626a4-75fc-403a-8016-4a1d2e3b034d.mp4

**Found in Branch**
development

**Commit ID from [o3de/o3de](https://github.com/o3de/o3de) Repository**
47eea0c820a3639fcbcb04aff991235e9132adb1 
Monday December 12 2022

### PR DESCRIPTION
**Pull request by hultonha**
## What does this PR do?

Fixes #13525
Fixes #13763

This PR is an incremental improvement and does not fix all issues with 'Be this camera', it does however make things dramatically better (see above issues). There are still several unfortunate problems with the behavior of the `ViewportCameraSelector` and `EditorCameraComponent`. After investigating it seemed like it was going to be a longer thing to investigate and fix (likely not happening in 2022) so I wanted to get this incremental PR posted to make things a bit better in the interim. I'll do my best to return to this in the new year along with more tests to verify the behavior.

## How was this PR tested?

Tested manually in the Editor. There are several tests that cover related functionality, more will be added in future.

Note: There are still problems when you go to Game Mode while still being in 'Be this camera', and there are problems when being in 'Be this camera' and opening a new level (see https://github.com/o3de/o3de/issues/13744).

Tagging [@Darking](https://github.com/Darkinggq) for visibility as well.

### CODE REVIEW

**Comment by hultonha** [comment]
Update: Did more testing and there still seems to be problems that need to be addressed. Will leave as-is for now. Please don't feel the need to review at this stage but you can test out the branch if you're interested.

File: Gems/Camera/Code/Source/CameraComponentController.cpp [hunk 1]
```diff
@@ -139,13 +139,7 @@ namespace Camera
 
     void CameraComponentController::DeactivateAtomView()
     {
-        if (!IsActiveView())
```
Comment (moudgils) [inline Gems/Camera/Code/Source/CameraComponentController.cpp:142]: Why remove this?

File: Gems/Atom/Tools/AtomToolsFramework/Code/Source/Viewport/ModularViewportCameraController.cpp [hunk 2]
```diff
@@ -381,7 +380,7 @@ namespace AtomToolsFramework
     {
         if (!m_storedCamera.has_value())
         {
-            m_storedCamera = m_previousCamera;
+            m_storedCamera = m_targetCamera;
         }
 
         const auto angles = AzFramework::EulerAngles(AZ::Matrix3x3::CreateFromQuaternion(worldFromLocal.GetRotation()));
```
Comment (rgba16f) [inline Gems/Atom/Tools/AtomToolsFramework/Code/Source/Viewport/ModularViewportCameraController.cpp:386]: The AZ Quaternion class has utility functions to retrieve the Euler angles (GetEulerDegrees, GetEulerRadians), no need to convert to a 3x3 matrix first.

File: Gems/Camera/Code/Source/CameraComponentController.cpp [hunk 3]
```diff
@@ -283,8 +277,7 @@ namespace Camera
         AZ::TransformNotificationBus::Handler::BusDisconnect(m_entityId);
         CameraRequestBus::Handler::BusDisconnect(m_entityId);
 
-        auto atomViewportRequests = AZ::Interface<AZ::RPI::ViewportContextRequestsInterface>::Get();
-        if (atomViewportRequests)
+        if (auto atomViewportRequests = AZ::Interface<AZ::RPI::ViewportContextRequestsInterface>::Get())
```
Comment (rgba16f) [inline Gems/Camera/Code/Source/CameraComponentController.cpp:280]: The `atomViewportRequests` variable isn't used, do you need to query the interface?

File: Gems/Camera/Code/Source/CameraComponentController.cpp (hunk 1 above)
Comment (hultonha) [reply Gems/Camera/Code/Source/CameraComponentController.cpp:142]: Sorry I meant to add a comment here on the review (I forgot as I moved this to draft for now as it needs more time)...

So this was actually leading to a crash in the `onChange` callback in the `Init` function of `CameraComponentController`. `onChange` captures `CameraComponentController` by value and assigns it to `m_atomCameraViewGroup`, which is a `shared_ptr`. Ownership of this is then shared by passing it to `PushViewGroup` (which stores the `shared_ptr`). When `CameraComponentController` gets destroyed and doesn't remove the reference to `m_atomCameraViewGroup`, it has a dangling pointer to the `CameraComponentController` instance which no longer exists, so we have to make sure to also remove it.

There might be a different way to resolve it but this felt like the most correct fix to me.

Comment (hultonha) [reply Gems/Camera/Code/Source/CameraComponentController.cpp:142]: Note: This crash only happens when you have more than one camera component in the scene and have one of the camera selected when moving to/from Game Mode

File: Gems/Atom/Tools/AtomToolsFramework/Code/Source/Viewport/ModularViewportCameraController.cpp (hunk 2 above)
Comment (hultonha) [reply Gems/Atom/Tools/AtomToolsFramework/Code/Source/Viewport/ModularViewportCameraController.cpp:386]: I am pretty sure I looked into this in in the past and it doesn't give the angles in the format I want (`AzFramework::EulerAngles` is specially written for the camera). This is outside the scope of the review for now so I'd rather leave as-is for now and I can check back on it as part of another PR if necessary.

File: Gems/Camera/Code/Source/CameraComponentController.cpp (hunk 3 above)
Comment (hultonha) [reply Gems/Camera/Code/Source/CameraComponentController.cpp:280]: Hmm yeah good point, I can just delete this entirely, the check before doesn't seem to be needed

File: Gems/Atom/Tools/AtomToolsFramework/Code/Source/Viewport/ModularViewportCameraController.cpp (hunk 2 above)
Comment (rgba16f) [reply Gems/Atom/Tools/AtomToolsFramework/Code/Source/Viewport/ModularViewportCameraController.cpp:386]: Seems fine to leave it for another PR. It's working, just not the most efficient.

**Review by moudgils — APPROVED** [review]

File: Gems/Atom/Tools/AtomToolsFramework/Code/Source/Viewport/ModularViewportCameraController.cpp (hunk 2 above)
Comment (hultonha) [reply Gems/Atom/Tools/AtomToolsFramework/Code/Source/Viewport/ModularViewportCameraController.cpp:386]: For sure, there are just so many versions of combining Euler angles the one used here is very specific (and handles us having Z as up too for example when converting to pitch/yaw/roll).

**Review by cgalvan — APPROVED** [review]

File: Gems/Atom/Tools/AtomToolsFramework/Code/Source/Viewport/ModularViewportCameraController.cpp [hunk 4]
```diff
@@ -172,13 +172,11 @@ namespace AtomToolsFramework
         controller->SetupCameraControllerPriority(m_priorityFn);
         controller->SetupCameraControllerViewportContext(m_modularCameraViewportContext);
 
-        auto handleCameraChange = [this]([[maybe_unused]] const AZ::Matrix4x4& cameraView)
+        auto handleCameraChangeFn = [this]([[maybe_unused]] const AZ::Matrix4x4& cameraView)
```
Comment (AMZN-Gene) [inline Gems/Atom/Tools/AtomToolsFramework/Code/Source/Viewport/ModularViewportCameraController.cpp:175]: The _Fn_ stood out as odd, but looks like it's the standard for this file (m_priorityFn).

**Review by AMZN-Gene — APPROVED** [review]

**Comment by AMZN-Gene** [comment]
Thanks @hultonha! Great quality of life fix :)

**Comment by hultonha** [comment]
> Thanks @hultonha! Great quality of life fix :)

No problem! Still working on a few more related fixes, hopefully should be able to finish this off next week 👍

File: Gems/Atom/Tools/AtomToolsFramework/Code/Source/Viewport/ModularViewportCameraController.cpp (hunk 4 above)
Comment (hultonha) [reply Gems/Atom/Tools/AtomToolsFramework/Code/Source/Viewport/ModularViewportCameraController.cpp:175]: That's a fair point... though we do use it in a few places, I'll have a think if it makes sense to keep or not, thanks for the suggestion! 👍

**Comment by hultonha** [comment]
This is being replaced by https://github.com/o3de/o3de/pull/14269 as there were lots more problems that were discovered...

**Pull request closed without merging**